### PR TITLE
[Testcase Refactoring] Decouple control flow CUDA initialization tests from hardware assumptions

### PR DESCRIPTION
**Pull request by ll-pen**
I reviewed the implementation and test results and confirm that the quoted description matches this change. Codex helped draft the quoted text.

> ## Summary
>
> Decouple `test_control_flow_cuda_initialization.py` from implicit hardware assumptions by adding explicit hardware classification labels.
>
> Part of #185590.
>
> ## Changes
>
> - Add corresponding `HardwareClassification` labels to test classes.
> - Preserve existing parameterization, backend admission, expected failures, and test behavior.
>
> ## Test Plan
>
> Ran:
>
> ```bash
> python -m py_compile test/functorch/test_control_flow_cuda_initialization.py
> lintrunner -a test/functorch/test_control_flow_cuda_initialization.py
> git diff --check
> ```
>
> CI:
>
> - Run the modified test file through its existing PyTorch test jobs.
> - Run the Inductor workflows selected by `ciflow/inductor`.

cc @wjlFlyer @pjyFight @FuDdd @fffrog @lyriexs666 @jishuangfeng @JamesD18 @JiasenTian

### CODE REVIEW

File: test/functorch/test_control_flow_cuda_initialization.py [hunk 1]
```diff
@@ -14,6 +15,8 @@
     "CUDA 12.4 or greater is required for CUDA Graphs with conditional nodes",
 )
 class TestControlFlowInCUDAGraphInitialization(TestCase):
+    hw_classification = HardwareClassification.CUDA
```
Comment (FuDdd) [inline test/functorch/test_control_flow_cuda_initialization.py:18]: Need to add instantiate_device_type_tests for this classification

Comment (ll-pen) [reply test/functorch/test_control_flow_cuda_initialization.py:18]: Thanks for pointing this out. I checked whether this class can safely use `instantiate_device_type_tests()`.

This test is unusual because it verifies CUDA initialization behavior from a process where the device primary context has not yet been created. `run_test.py` therefore runs each test in a separate subprocess, and `setUp()` asserts that `torch._C._cuda_hasPrimaryContext(0)` is still false.

 `instantiate_device_type_tests(..., only_for=("cuda",))` would generate a class derived from `CUDATestBase`. Before the per-test `setUp()` runs, `CUDATestBase.setUpClass()` executes:

```python
t = torch.ones(1).cuda()
```

That creates the CUDA primary context before this test can perform its initial assertion, causing the test to fail or invalidating the initialization behavior it is intended to cover. Changing the class-name reference in `setUp()` to `self.CTX_ALREADY_CREATED_ERR_MSG` would avoid a `NameError`, but would not address this earlier context creation.

For that reason, I believe this class needs to remain a direct `TestCase` compatibility exception, with `HardwareClassification.CUDA` providing the classification metadata and the existing `TEST_CUDA_GRAPH_CONDITIONAL_NODES` class-level skip providing runtime admission. Would you agree with keeping it uninstantiated for this reason?

**Review by FuDdd — APPROVED** [review]

**Review by KarhouTam — COMMENTED** [review]
No `instantiate_device_type_tests`?

**Comment by ll-pen** [comment]
> No `instantiate_device_type_tests`?

Thanks for pointing this out. I checked whether this class can safely use instantiate_device_type_tests().

This test is unusual because it verifies CUDA initialization behavior from a process where the device primary context has not yet been created. run_test.py therefore runs each test in a separate subprocess, and setUp() asserts that torch._C._cuda_hasPrimaryContext(0) is still false.

instantiate_device_type_tests(..., only_for=("cuda",)) would generate a class derived from CUDATestBase. Before the per-test setUp() runs, CUDATestBase.setUpClass() executes:

t = torch.ones(1).cuda()
That creates the CUDA primary context before this test can perform its initial assertion, causing the test to fail or invalidating the initialization behavior it is intended to cover. Changing the class-name reference in setUp() to self.CTX_ALREADY_CREATED_ERR_MSG would avoid a NameError, but would not address this earlier context creation.

For that reason, I believe this class needs to remain a direct TestCase compatibility exception, with HardwareClassification.CUDA providing the classification metadata and the existing TEST_CUDA_GRAPH_CONDITIONAL_NODES class-level skip providing runtime admission. Would you agree with keeping it uninstantiated for this reason?

**Comment by ll-pen** [comment]
Hi @zou3519 , validation is complete. Could you take a look and help merge this PR? Thanks!

**Review by KarhouTam — APPROVED** [review]
cc @fffrog

**Review by fffrog — CHANGES_REQUESTED** [review]
Please follow the principle of refactoring testcases to make it to be `instantiate` form.

**Review by orangeH25 — APPROVED** [review]

**Review by fffrog — APPROVED** [review]
SGTM

Hi @ll-pen, this test file is a bit unusual, so I think adding more description in line 18 to explain why we do not refactor it into `instantiate` form should be better.

**Comment by ll-pen** [comment]
> SGTM
> 
> Hi @ll-pen, this test file is a bit unusual, so I think adding more description in line 18 to explain why we do not refactor it into `instantiate` form should be better.

Done.
